src/storage/engines/FirebaseStorageEngine.ts

```diff
@@ -555,32 +499,257 @@ export class FirebaseStorageEngine extends StorageEngine {
+    await this._copyDirectory(`${snapshotName}/participants`, `${originalName}/participants`);
+    await this._copyDirectory(snapshotName, originalName);
+    await this._copyCollection(snapshotName, originalName);
+  }
+
+  // Function to add collection name to metadata
+  async _addDirectoryNameToMetadata(directoryName: string) {
+    try {
+      const metadataDoc = doc(this.firestore, 'metadata', 'collections');
+      await setDoc(metadataDoc, { [directoryName]: true }, { merge: true });
+    } catch (error) {
+      console.error('Error adding collection to metadata:', error);
+    }
+  }
+
+  async _removeNameFromMetadata(directoryName: string) {
+    try {
+      const metadataDoc = doc(this.firestore, 'metadata', 'collections');
+      const metadataSnapshot = await getDoc(metadataDoc);
+
+      if (metadataSnapshot.exists()) {
+        const metadata = metadataSnapshot.data();
+        if (metadata[directoryName]) {
+          delete metadata[directoryName];
+          await setDoc(metadataDoc, metadata);
+        } else {
+          console.warn(`${directoryName} does not exist in metadata.`);
+        }
+      } else {
+        console.warn('No metadata found.');
+      }
+    } catch (error) {
+      console.error('Error removing collection from metadata:', error);
+      throw error;
+    }
+  }
+
+  async _deleteDirectory(directoryPath: string) {
+    try {
+      const directoryRef = ref(this.storage, directoryPath);
+      const directorySnapshot = await listAll(directoryRef);
+
+      const deletePromises = directorySnapshot.items.map((fileRef) => deleteObject(fileRef));
+      await Promise.all(deletePromises);
+    } catch (error) {
+      console.error('Error deleting files in directory:', error);
+    }
+  }
+
+  async _copyDirectory(sourceDir: string, targetDir: string) {
+    try {
+      const sourceDirRef = ref(this.storage, sourceDir);
+      const sourceDirSnapshot = await listAll(sourceDirRef);
+
+      const copyPromises = sourceDirSnapshot.items.map((fileRef) => {
+        const sourceFilePath = fileRef.fullPath;
+        const targetFilePath = sourceFilePath.replace(sourceDir, targetDir);
+        return this._copyFile(sourceFilePath, targetFilePath);
+      });
+      await Promise.all(copyPromises);
+    } catch (error) {
+      console.error('Error copying file:', error);
+    }
+  }
+
+  async _copyFile(sourceFilePath: string, targetFilePath: string) {
+    try {
+      const sourceFileRef = ref(this.storage, sourceFilePath);
+      const targetFileRef = ref(this.storage, targetFilePath);
+
+      const sourceFileURL = await getDownloadURL(sourceFileRef);
+      const response = await fetch(sourceFileURL);
+      const blob = await response.blob();
+
+      await uploadBytes(targetFileRef, blob);
+    } catch (error) {
+      console.error('Error copying file:', error);
+    }
+  }
+
+  async _copyCollection(sourceCollectionName: string, targetCollectionName: string) {
```

Comment:
These `_{copy,delete}collection` methods are going to sting us when we add anything else to the firestore... I think we need to make these more general, but I don't want to slow this PR down. Let's make an issue for these for when we have some extra cycles

Reply:
Yeah, I agree. It was simpler to do this rather than make a recursive one with all these edge cases since we really only copy this. I'll make an issue for it